Assignment detail crash

Steps to reproduce

1. Open app
2. Make sure you have soldier
3. Open assignments
4. Scroll down to single mission assignments that are at the end
5. Tap to open single mission assignment to see details
6. App crashes

````
Process:com.ninetwozero.bf4intel,PID:28951
        java.lang.RuntimeException:Unable to start activity ComponentInfo{com.ninetwozero.bf4intel/com.ninetwozero.bf4intel.ui.activities.SingleFragmentActivity}:java.lang.IndexOutOfBoundsException:Invalid index 0,size is 0
        at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:2416)
        at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:2476)
        at android.app.ActivityThread.-wrap11(ActivityThread.java)
        at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1344)
        at android.os.Handler.dispatchMessage(Handler.java:102)
        at android.os.Looper.loop(Looper.java:148)
        at android.app.ActivityThread.main(ActivityThread.java:5417)
        at java.lang.reflect.Method.invoke(Native Method)
        at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:726)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:616)
        Caused by:java.lang.IndexOutOfBoundsException:Invalid index 0,size is 0
        at java.util.ArrayList.throwIndexOutOfBoundsException(ArrayList.java:255)
        at java.util.ArrayList.get(ArrayList.java:308)
        at com.ninetwozero.bf4intel.ui.assignments.AssignmentDetailFragment.populateAwardRequirements(AssignmentDetailFragment.java:185)
        at com.ninetwozero.bf4intel.ui.assignments.AssignmentDetailFragment.populateViews(AssignmentDetailFragment.java:90)
        at com.ninetwozero.bf4intel.ui.assignments.AssignmentDetailFragment.initialize(AssignmentDetailFragment.java:70)
        at com.ninetwozero.bf4intel.ui.assignments.AssignmentDetailFragment.onCreateView(AssignmentDetailFragment.java:64)
        at android.support.v4.app.Fragment.performCreateView(Fragment.java:2248)
        at android.support.v4.app.FragmentManagerImpl.moveToState(FragmentManager.java:1340)
        at android.support.v4.app.FragmentManagerImpl.moveFragmentToExpectedState(FragmentManager.java:1569)
        at android.support.v4.app.FragmentManagerImpl.moveToState(FragmentManager.java:1636)
        at android.support.v4.app.BackStackRecord.executeOps(BackStackRecord.java:758)
        at android.support.v4.app.FragmentManagerImpl.executeOps(FragmentManager.java:2415)
        at android.support.v4.app.FragmentManagerImpl.executeOpsTogether(FragmentManager.java:2201)
        at android.support.v4.app.FragmentManagerImpl.optimizeAndExecuteOps(FragmentManager.java:2155)
        at android.support.v4.app.FragmentManagerImpl.execPendingActions(FragmentManager.java:2064)
        at android.support.v4.app.FragmentController.execPendingActions(FragmentController.java:379)
        at android.support.v4.app.FragmentActivity.onStart(FragmentActivity.java:607)
        at android.support.v7.app.AppCompatActivity.onStart(AppCompatActivity.java:178)
        at android.app.Instrumentation.callActivityOnStart(Instrumentation.java:1237)
        at android.app.Activity.performStart(Activity.java:6268)
        at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:2379)
        at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:2476) 
        at android.app.ActivityThread.-wrap11(ActivityThread.java) 
        at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1344) 
        at android.os.Handler.dispatchMessage(Handler.java:102) 
        at android.os.Looper.loop(Looper.java:148) 
        at android.app.ActivityThread.main(ActivityThread.java:5417) 
        at java.lang.reflect.Method.invoke(Native Method) 
        at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:726) 
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:616)
````

Thought based on stacktrace - the assignment doesn't have any
prerequisites? 🤔

On Fri, Jan 5, 2018, 09:53 Peter Mikloško <notifications@github.com> wrote:

> Steps to reproduce
>
>    1. Open app
>    2. Make sure you have soldier
>    3. Open assignments
>    4. Scroll down to single mission assignments that are at the end
>    5. Tap to open single mission assignment to see details
>    6. App crashes
>
> Process: com.ninetwozero.bf4intel, PID: 28951
>                                                    java.lang.RuntimeException: Unable to start activity ComponentInfo{com.ninetwozero.bf4intel/com.ninetwozero.bf4intel.ui.activities.SingleFragmentActivity}: java.lang.IndexOutOfBoundsException: Invalid index 0, size is 0
>                                                        at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:2416)
>                                                        at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:2476)
>                                                        at android.app.ActivityThread.-wrap11(ActivityThread.java)
>                                                        at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1344)
>                                                        at android.os.Handler.dispatchMessage(Handler.java:102)
>                                                        at android.os.Looper.loop(Looper.java:148)
>                                                        at android.app.ActivityThread.main(ActivityThread.java:5417)
>                                                        at java.lang.reflect.Method.invoke(Native Method)
>                                                        at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:726)
>                                                        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:616)
>                                                     Caused by: java.lang.IndexOutOfBoundsException: Invalid index 0, size is 0
>                                                        at java.util.ArrayList.throwIndexOutOfBoundsException(ArrayList.java:255)
>                                                        at java.util.ArrayList.get(ArrayList.java:308)
>                                                        at com.ninetwozero.bf4intel.ui.assignments.AssignmentDetailFragment.populateAwardRequirements(AssignmentDetailFragment.java:185)
>                                                        at com.ninetwozero.bf4intel.ui.assignments.AssignmentDetailFragment.populateViews(AssignmentDetailFragment.java:90)
>                                                        at com.ninetwozero.bf4intel.ui.assignments.AssignmentDetailFragment.initialize(AssignmentDetailFragment.java:70)
>                                                        at com.ninetwozero.bf4intel.ui.assignments.AssignmentDetailFragment.onCreateView(AssignmentDetailFragment.java:64)
>                                                        at android.support.v4.app.Fragment.performCreateView(Fragment.java:2248)
>                                                        at android.support.v4.app.FragmentManagerImpl.moveToState(FragmentManager.java:1340)
>                                                        at android.support.v4.app.FragmentManagerImpl.moveFragmentToExpectedState(FragmentManager.java:1569)
>                                                        at android.support.v4.app.FragmentManagerImpl.moveToState(FragmentManager.java:1636)
>                                                        at android.support.v4.app.BackStackRecord.executeOps(BackStackRecord.java:758)
>                                                        at android.support.v4.app.FragmentManagerImpl.executeOps(FragmentManager.java:2415)
>                                                        at android.support.v4.app.FragmentManagerImpl.executeOpsTogether(FragmentManager.java:2201)
>                                                        at android.support.v4.app.FragmentManagerImpl.optimizeAndExecuteOps(FragmentManager.java:2155)
>                                                        at android.support.v4.app.FragmentManagerImpl.execPendingActions(FragmentManager.java:2064)
>                                                        at android.support.v4.app.FragmentController.execPendingActions(FragmentController.java:379)
>                                                        at android.support.v4.app.FragmentActivity.onStart(FragmentActivity.java:607)
>                                                        at android.support.v7.app.AppCompatActivity.onStart(AppCompatActivity.java:178)
>                                                        at android.app.Instrumentation.callActivityOnStart(Instrumentation.java:1237)
>                                                        at android.app.Activity.performStart(Activity.java:6268)
>                                                        at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:2379)
>                                                        at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:2476)
>                                                        at android.app.ActivityThread.-wrap11(ActivityThread.java)
>                                                        at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1344)
>                                                        at android.os.Handler.dispatchMessage(Handler.java:102)
>                                                        at android.os.Looper.loop(Looper.java:148)
>                                                        at android.app.ActivityThread.main(ActivityThread.java:5417)
>                                                        at java.lang.reflect.Method.invoke(Native Method)
>                                                        at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:726)
>                                                        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:616)
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/ninetwozero/com.ninetwozero.bf4intel/issues/189>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AA7w0GmHTXNmb54r6MhWu2yy2Ai5HysLks5tHeMYgaJpZM4RULmd>
> .
>

--- a/BF4Intel/build.gradle
+++ b/BF4Intel/build.gradle
@@ -17,8 +17,8 @@ android {
 
     defaultConfig {
         applicationId 'com.ninetwozero.bf4intel'
-        versionCode 1614120
-        versionName '1.2.0'
+        versionCode 1614121
+        versionName '1.2.1'
         minSdkVersion 14
         targetSdkVersion 25
     }

--- a/BF4Intel/src/main/assets/changelog.htm
+++ b/BF4Intel/src/main/assets/changelog.htm
@@ -41,6 +41,14 @@
     </style>
 </head>
 <body>
+<h1 class="group-header">Version 1.2.1</h1>
+<article>
+    <ul>
+        <li>Fixed crash when user try to view assignment details of single player assignments</li>
+        <li>Fixed crash if user viewing AWARDS screen tap on sort of filter icon before data are loaded on screen</li>
+        <li>Fixed crash when viewing NEWS and server doesn't returns any results. User will is presented with message to try later</li>
+    </ul>
+</article>
 <h1 class="group-header">Version 1.2.0</h1>
 <article>
     <ul>

--- a/BF4Intel/src/main/java/com/ninetwozero/bf4intel/ui/assignments/AssignmentDetailFragment.java
+++ b/BF4Intel/src/main/java/com/ninetwozero/bf4intel/ui/assignments/AssignmentDetailFragment.java
@@ -181,7 +181,7 @@ public class AssignmentDetailFragment extends BaseDialogFragment {
             final boolean isInRoundRequirement = "CriteriaType_IAR_InARound".equals(criteria.getCriteriaType());
 
             setText(tempView, R.id.task_label, AssignmentCriteriaStringMap.get(criteria.getKey()));
-            if(assignment.getAward().getAwardCriterias() != null) { //User updated data nad have new AwardsCriterias object
+            if(assignment.getAward().getAwardCriterias() != null && !assignment.getAward().getAwardCriterias().isEmpty()) { //User updated data nad have new AwardsCriterias object
                 setText(tempView, R.id.task_completion, getTaskCompletionString(criteria.getCurrentValue(), assignment.getAward().getAwardCriterias().get(i).getCompletionValue()));
             } else {//User is still on old data so use old format
                 setText(tempView, R.id.task_completion, getTaskCompletionString(criteria));

--- a/BF4Intel/src/main/java/com/ninetwozero/bf4intel/ui/awards/AwardGridFragment.java
+++ b/BF4Intel/src/main/java/com/ninetwozero/bf4intel/ui/awards/AwardGridFragment.java
@@ -121,14 +121,16 @@ public class AwardGridFragment
         final View emptyView = view.findViewById(android.R.id.empty);
         setCustomEmptyText(emptyView, R.string.empty_text_awards);
 
-        final GridView gridView = (GridView) view.findViewById(R.id.assignments_grid);
+        gridView = (GridView) view.findViewById(R.id.assignments_grid);
         gridView.setOnItemClickListener(this);
         gridView.setEmptyView(emptyView);
 
     }
 
     private void sendDataToGridView(final View view, List<Award> awards) {
-        gridView = (GridView) view.findViewById(R.id.assignments_grid);
+        if(gridView == null) {
+            gridView = (GridView) view.findViewById(R.id.assignments_grid);
+        }
         AwardsAdapter adapter = (AwardsAdapter) gridView.getAdapter();
         if (adapter == null) {
             adapter = new AwardsAdapter(getActivity());

--- a/BF4Intel/src/main/java/com/ninetwozero/bf4intel/ui/news/NewsListingFragment.java
+++ b/BF4Intel/src/main/java/com/ninetwozero/bf4intel/ui/news/NewsListingFragment.java
@@ -6,6 +6,7 @@ import android.view.LayoutInflater;
 import android.view.View;
 import android.view.ViewGroup;
 import android.widget.ListView;
+import android.widget.Toast;
 
 import com.android.volley.VolleyError;
 import com.ninetwozero.bf4intel.Bf4Intel;
@@ -21,6 +22,8 @@ import com.ninetwozero.bf4intel.ui.menu.RefreshEvent;
 import com.squareup.otto.Subscribe;
 
 import java.util.List;
+
+import javax.xml.datatype.Duration;
 
 public class NewsListingFragment extends BaseLoadingListFragment {
     public static final String ID = "articleId";
@@ -84,7 +87,11 @@ public class NewsListingFragment extends BaseLoadingListFragment {
     @Subscribe
     @SuppressWarnings("unused")
     public void onNewsRefreshed(NewsListingRefreshedEvent response) {
-        sendItemsToListView(response.getRequest().getArticles());
+        if (response.getRequest() != null && response.getRequest().getArticles() != null) {
+            sendItemsToListView(response.getRequest().getArticles());
+        } else {
+            Toast.makeText(getContext(), "Failed to load latest new, please try later.", Toast.LENGTH_LONG).show();
+        }
         showLoadingState(false);
         isReloading = false;
     }

--- a/BF4Intel/src/main/java/com/ninetwozero/bf4intel/ui/news/NewsListingFragment.java
+++ b/BF4Intel/src/main/java/com/ninetwozero/bf4intel/ui/news/NewsListingFragment.java
@@ -23,8 +23,6 @@ import com.squareup.otto.Subscribe;
 
 import java.util.List;
 
-import javax.xml.datatype.Duration;
-
 public class NewsListingFragment extends BaseLoadingListFragment {
     public static final String ID = "articleId";
 

